src/shell/Shell.zig

```diff
@@ -0,0 +1,137 @@
+const ft = @import("../ft/ft.zig");
+const token = @import("token.zig");
+const utils = @import("default/utils.zig");
+const allocator = @import("../memory.zig").physicalMemory.allocator();
+
+pub const CmdError = error{ CommandNotFound, InvalidNumberOfArguments, InvalidParameter, OtherError };
+pub const Config = struct {
+    colors: bool = true,
+};
+
+const max_line_size: u16 = 1024;
+
+pub fn Shell(comptime _builtins: anytype) type {
+    return struct {
+        const builtins = _builtins;
+        const Self = @This();
+
+        const Hook: type = ?*const fn (*Self) void;
+        const ErrorHook: type = ?*const fn (*Self, err: anyerror) void;
+
+        const Hooks: type = struct {
+            pre_process: Hook = null,
+            post_process: Hook = null,
+            pre_cmd: Hook = null,
+            on_error: ErrorHook = null,
+        };
+
+        config: Config = Config{},
+        err: bool = false,
+        reader: ft.io.AnyReader = undefined,
+        writer: ft.io.AnyWriter = undefined,
+        hooks: Hooks = Hooks{},
+
+        pub fn init(
+            reader: ft.io.AnyReader,
+            writer: ft.io.AnyWriter,
+            config: Config,
+            hooks: struct {
+                pre_process: ?*const anyopaque = null,
+                post_process: ?*const anyopaque = null,
+                pre_cmd: ?*const anyopaque = null,
+                on_error: ?*const anyopaque = null,
```

Comment:
why can´t hooks be of type Hooks?

Reply:
Circular dependencies are a real pain... :upside_down_face:

Reply:
I don't remember exactly why, but I had a lot of errors about circular dependencies, and if I wanted to use the Hook type, I had to declare init as a comptime-known method (and all it args too...) or something like that, and I don't really like that